### DISCUSSION TITLE
Custom Login

### DISCUSSION THREAD

**Discussion by ryzam** [Budibase/budibase#3214] [Help]
Hi guys, any example of how I can do custom login using JWT token?
![image](https://user-images.githubusercontent.com/597759/139673894-b30275fb-8c54-49af-83b5-13d6d437da64.png)

**Comment by mjashanks** [comment]
This is not possible. Is there a certain SSO provider/mechanism that you would like us to support for the main login screen?

**Reply by ryzam** [reply]
I would like to authenticate users using Headless CMS Strapi that will return JWT token when we submit username and password

**Reply by mjashanks** [reply]
Thanks. Yea, there is no way to do this today and it won't be possible anytime soon. 

If we were going to offer anything like this, we would need to make authentication pluggable - thus allowing people to write their own auth for Budibase. That would be a great feature to have, but low on the priority list right now.

**Reply by evtnl** [reply]
This would be REALLY nice to have! I have my own SSO/user management system and would love to be able using that.

**Reply by rwagnergit** [reply]
I'll echo evtnl.  I just spent a couple days looking at budibase.  Love it.  But without custom authentication, it's a non starter for us.  We have an internal REST API that takes username/password and returns a JWT.  We need budibase to capture that and use it in subsequent REST calls (to our REST API datasources).

**Reply by evtnl** [reply]
what rwagnergit suggests is a really great addition, though it is not authentication to Budibase, but from Budibase to external API's. Normally you would use key-based auth for this, and add the user to the parameters of the call. 
Using the JWT would be a valid addition. However, we would probably need to be able to use custom user fields, which is not possible because of bug [7368](https://github.com/Budibase/budibase/issues/7368)

**Comment by eftomi** [comment]
I tried to implement a pretty rudimentary JWT authorization scheme to access some backend data service which uses JWT through Budibase REST API queries, not for the actual Budibase login. I prepared two REST API queries, the first one to create a user in backend service, and the second one to request the token which would be consequently called if the token expires. I saved the token and Base64 usrn&pwd combination into the Users table. However, I was not successful in doing that, because custom columns in Users table cannot be directly used in REST API queries:

https://github.com/Budibase/budibase/discussions/8824

Then I tried with the idea to save the token as a state - I run the query to fetch the token on each screen load as a screen action, passing the auth. data from Users table, however during the actual screen load I get "Cannot read property 'user' of undefined" error, then the token is received, and then another "Cannot read property 'user' of undefined" error happens.
